app/models/chess_piece.rb

```diff
@@ -9,4 +9,61 @@ class ChessPiece < ActiveRecord::Base
             :color,
             presence: true
   enum color: [:black, :white]
+
+  def obstructed?(destination_x, destination_y)
+    if diagonal_move?(destination_x, destination_y)
+      diagonal_obstruction?(destination_x, destination_y)
+    elsif vertical_move?(destination_x, destination_y)
+      vertical_obstruction?(destination_x, destination_y)
+    elsif horizontal_move?(destination_x, destination_y)
+      horizontal_obstruction?(destination_x, destination_y)
+    else
+      fail ArgumentError, 'Invalid move'
+    end
+  end
+
+  private
+
+  def diagonal_obstruction?(destination_x, destination_y)
+    from_x, to_x = [position_x, destination_x].minmax
+    from_y = [position_y, destination_y].min
+    ((from_x + 1)...to_x).each_with_index do |x, idx|
+      return true if position_occupied?(x, from_y + idx + 1)
+    end
+    false
```

Comment:
I'm thinking that since from_x is only used as `from_x + 1`. 

This could make it more readable.

```Ruby
from_x = [position_x, destination_x].min + 1
to_x = [position_x, destination_x].max
from_y = [position_y, destination_y].min

from_x...to_x.each_with_index do |x, idx|
  return true if position_occupied?(x, from_y + idx + 1)
end

false
```